market_maker_keeper/leverj_market_maker_keeper.py

```diff
@@ -0,0 +1,244 @@
+            self.web3 = Web3(HTTPProvider(endpoint_uri=f"http://{self.arguments.rpc_host}:{self.arguments.rpc_port}",
+                                      request_kwargs={"timeout": self.arguments.rpc_timeout}))
+
+        self.web3.eth.defaultAccount = self.arguments.eth_from
+        register_keys(self.web3, self.arguments.eth_key)
+
+        setup_logging(self.arguments)
+
+        self.bands_config = ReloadableConfig(self.arguments.config)
+        self.price_feed = PriceFeedFactory().create_price_feed(self.arguments)
+        self.spread_feed = create_spread_feed(self.arguments)
+        self.control_feed = create_control_feed(self.arguments)
+        self.order_history_reporter = create_order_history_reporter(self.arguments)
+
+        self.history = History()
+
+        self.leverj_api = LeverjAPI(web3=self.web3,
+                                    api_server=self.arguments.leverj_api_server,
+                                    account_id=self.arguments.account_id,
+                                    api_key=self.arguments.api_key,
+                                    api_secret=self.arguments.api_secret,
+                                    timeout=self.arguments.leverj_timeout)
+
+
+        self.order_book_manager = OrderBookManager(refresh_frequency=self.arguments.refresh_frequency)
+        self.order_book_manager.get_orders_with(lambda: self.leverj_api.get_orders(self.pair()))
+        self.order_book_manager.get_balances_with(lambda: self.leverj_api.get_balances())
+        self.order_book_manager.cancel_orders_with(lambda order: self.leverj_api.cancel_order(order.order_id))
+        self.order_book_manager.enable_history_reporting(self.order_history_reporter, self.our_buy_orders,
+                                                         self.our_sell_orders)
+        self.order_book_manager.start()
+
+    def main(self):
+        with Lifecycle() as lifecycle:
+            lifecycle.initial_delay(1)
+            lifecycle.on_startup(self.startup)
+            lifecycle.every(1, self.synchronize_orders)
+            lifecycle.on_shutdown(self.shutdown)
+
+    def startup(self):
+        quote_increment = 1/(self.leverj_api.get_product(self.arguments.pair)["ticksperpoint"])
+        self.precision = -(int(log10(float(quote_increment)))+1)
+
+    def shutdown(self):
+        self.order_book_manager.cancel_all_orders()
+
+    def pair(self):
+        return self.arguments.pair.upper()
+
+    def token_sell(self) -> str:
+        if self.arguments.pair.startswith("USDC"):
+            return "USDC"
+        return self.arguments.pair[:3]
+
+    def token_buy(self) -> str:
+        if self.arguments.pair.startswith("USDC"):
+            return self.arguments.pair[4:]
+        return self.arguments.pair[3:]
+
+
+    def our_available_balance(self, our_balances: dict, token: str) -> Wad:
+        for key in our_balances:
+            if our_balances[key]['symbol'] == token:
+                if (token == "LEV") or (token == "FEE"): 
```

Comment:
Does the exchange provide metadata describing the number of decimal places for each token?  Seems dangerous to hardcode this.

Reply:
No, I think it's dangerous to rely on exchange meta data for this. If the exchange screws up this very basic information, we will be the ones that lose money on their error. It's better to hard code this or get this data from our side. On that note , if model.py was approved in pymaker, I could probably use that and our configs to pass in this data.

Reply:
Let's go with hardcoded, this is info that you can easily screw up with configs, and the consequences are large if you mess up in the config